lunabot_nav/scripts/dstar_node.py

```diff
@@ -9,157 +9,224 @@
+                temp_map[i][j] = data.data[index]
+                index += 1
 
-    odom_topic = rospy.get_param("/odom_topic")
-    goal_topic = rospy.get_param("/nav_goal_topic")
+        self.grid = temp_map.copy()
+        self.grid_update_needed = True
 
-    radius = 8  # robot rad (grid units)
+    def convert_to_grid(self, position: 'list[float]') -> 'list[int]':
+        """
+        Convert a real world (x y) position to grid coordinates. Grid offset should be in the same frame as position, the grid is row-major.
+        """
 
-    frequency = 10  # hz
+        shifted_pos = [position[0] - self.x_offset, position[1] - self.y_offset]
+        coord = [
+            int(shifted_pos[1] / self.resolution + 0.5),
+            int(shifted_pos[0] / self.resolution + 0.5),
+        ]
+        return coord
 
-    dstar = None
+    
+    def adjust_for_goal(self):
+        """
+        Adjust the map (expand if needed) to adjust for a goal that is beyond the range of the map
+        """
 
-    path_publisher = rospy.Publisher("/nav/global_path", Path, queue_size=10)
+        if len(self.goal) == 0:
+            return
+        
+        goal_grid_coords = self.convert_to_grid(self.goal)
 
-    rospy.Subscriber(
-        "/maps/costmap_node/global_costmap/costmap", OccupancyGrid, grid_subscriber
-    )
+        if (goal_grid_coords[0] < 0):
+            # expand map upwards
+            difference = abs(goal_grid_coords[0])
 
-    rospy.Subscriber(
-        "/maps/costmap_node/global_costmap/costmap_updates",
-        OccupancyGridUpdate,
-        grid_update_subscriber,
-    )
+            rows = np.ones((difference, len(self.grid[0]))) * -1
+            self.grid = np.concatenate((rows, self.grid), axis=0)
 
-    rospy.Subscriber(odom_topic, Odometry, position_subscriber)
+            # change offset
+            self.y_offset -= difference * self.resolution
 
-    rospy.Subscriber(goal_topic, PoseStamped, goal_subscriber)
+        elif (goal_grid_coords[0] >= len(self.grid)):
+            # expand map downwards
+            difference = goal_grid_coords[0] - (len(self.grid) - 1)
 
-    rate = rospy.Rate(frequency)
+            rows = np.ones((difference, len(self.grid[0]))) * -1
+            self.grid = np.concatenate((self.grid, rows), axis=0)
 
-    completedInitialRun = False
+        if (goal_grid_coords[1] < 0):
+            # expand map to the left
+            
+            difference = abs(goal_grid_coords[1])
 
-    path = []
-    while not rospy.is_shutdown():
-        # print("Loop")
-        if (dstar is None) and len(grid) > 0 and len(pose) > 0 and len(goal) > 0:
-            dstar = Dstar(goal, pose, grid, radius, res, x_offset, y_offset)
-            print("Initialize")
+            columns = np.ones((len(self.grid), difference)) * -1
+            self.grid = np.concatenate((columns, self.grid), axis=1)
 
-        if dstar is not None:
-            # print("Iterate")
+            # change offset
```

Comment:
How is this logic handled when a new map update occurs? I don't see any logic about x_offset being fixed when updating map so will it override the wrong cells in the map?

Reply:
Same issue for `y_axis` code update above

Reply:
Yes this is the problem